### ISSUE TITLE
We should be able to select which version of ansible gets set up in the build container

### ISSUE THREAD

**Issue by dmsimard** [ansible/ansible-container#150]
Right now the ansible version is hardcoded in the build container: https://github.com/ansible/ansible-container/blob/707370bd7f02173be97abd4c19f94d166b146683/container/templates/ansible-dockerfile.j2#L14

Users, for any number of reasons, should be able to select the version of Ansible they'll be installing in the build container. For example when they need a new version, or pin to an older version due to backwards compatibility break and what have you.

I naively expected ansible-container to use whatever version of Ansible I had currently installed locally.

**Comment by dmsimard** [comment]
I guess there should be a mean to customize the build container in general. For example if I want to install callbacks, plugins or custom modules.

**Comment by chouseknecht** [comment]
This is similar to #151. 

@j00bar It seems there is a desire to customize the --local-build option, meaning that when running a local build users should have more control over the Dockerfile used to create the Ansible build container.

Would it make sense to support an *ansible-dockerfile.j2* in the ansible directory? If ansible-container finds *ansible/ansible-dockerfile.j2*, then use it over the default. This way users could change the version of Ansible and possibly add other customizations.

**Comment by dmsimard** [comment]
@chouseknecht yeah, the customization of the build container could potentially go beyond just selecting what version of Ansible to build with. For example to add third party modules, plugins/callbacks, etc.

**Comment by j00bar** [comment]
For the Ansible version, since the Dockerfile is templated already, let's support another commandline option for `build` when using `--local-builder` for `--ansible-version`. If the value given with this argument is "develop", the local builder should install the latest Ansible from GitHub with `-e` passed to `pip`. If not, it should use the version value given. Default should always go to the latest Ansible release.

**Comment by dmsimard** [comment]
@j00bar this might go back to a fundamental architecture decision and my objective is not to challenge that but I want to understand.
Why does ansible-container use a build container (in which Ansible is installed) at all ? Why couldn't it use whatever version of Ansible is installed locally and use the docker modules to start, connect and manage the containers ?

**Comment by j00bar** [comment]
@dmsimard Totally understand where you're coming from here, but part of the desired/expected direction of this project will be toward non-Ansible-veterans and even more so non-*NIX users. Shipping Ansible as part of a container lowers the barrier to entry for this demographic.

**Comment by dmsimard** [comment]
@j00bar Sorry for sidetracking away from the original issue but I feel this is somewhat related.
I'm not convinced about your comment... (not that you have to convince me, but take that as feedback).

The end user will have to install ansible-container to use it, anyway, right ?
Ansible could be a [requirement in ansible-container](https://github.com/ansible/ansible-container/blob/develop/requirements.txt) with a constraint set to ensure that things work (i.e, ansible >= 2.1.0.0).

By having Ansible as a requirement, it will automatically be installed when ansible-container is installed. If I happen to have Ansible 2.2 because that's what I'm working with, ansible-container will see that the requirement is met and not touch my local installation.
If I have callbacks, plugins, or custom ansible modules installed, they'll work out of the box.

By shipping Ansible in a container, we're just making it awkward to customize that particular Ansible installation (as shown here) -- and worse, the developers of ansible-container will have to maintain and develop that container.
People do not only care about the version of Ansible, they might require additional plugins, modules, callbacks -- so you'll eventually have to add support for hooks to install whatever in there. It'll end up being more complicated, IMHO.

**Comment by chouseknecht** [comment]
@dmsimard I agree that the local installation of Ansible can become very customized. But the idea of maintaining that locally and having ansible-container rely on the local install feels like the wrong strategy, IMHO. To me the question is, if there exists in my environment a complex install of Ansible, how do I easily replicate it throughout the organization and across dev, test and production environments? Putting it into a container image seems like a good solution. If you have an Ansible image with the right version, plugins, modules, etc. it would be easy to distribute and it should be easy to plug into ansible-container. That seems more valuable than relying on a local install of Ansible.

Seems like ansible-container should provide the ability to plug in a custom build image. It povides the ability to twist a few knobs via --with-variables, --with-volumes, --local-builder and --ansible-version to control what it creates as a build image, but why not just let the user have full control by bringing there own build image?

**Comment by containscafeine** [comment]
I had a similar discussion with my peers here and one does imply that the local install of ansible will be used to provision the container images, however not adding it to `requirements.txt` does make sense for a larger audience.

Does it work if we were to give the user an option to use the local install of ansible instead of using the builder container; or maybe just using the local install if available, and if not, get the builder container and use that instead?
Also, +1 for bringing your own customized ansible container image.

Does this make sense?

**Comment by j00bar** [comment]
Given the desire to make Ansible Container as easy as possible for, say, the Java on Windows in Eclipse Ché user, we're going to continue with the build container approach. I'm +1 on the idea of bringing your own build container and on being able to specify the Ansible version to customize a local builder.

**Comment by dmsimard** [comment]
Yeah, I think it's a good compromise to provide the ability to use your own builld container.
It would simultaneously allow for the customization users would want/need while also keeping upstream maintenance of the "default" build container low-effort.

**Comment by gregdek** [comment]
Changing the title of this issue to reflect the new consensus on what it should do.
